### PR TITLE
dockerclient: Agent supported Docker versions from client min,API Versions

### PR DESCRIPTION
**Pull request by sharanyad**
### Summary
Finds Agent supported Docker version clients from client version's min, API Versions returned

### Implementation details
dockerclient: Finding agent supported Docker versions by getting the default client version and if the min API version is present, add all the clients between min API version and API version supported, else follow current logic
utils: reuse the semantic version comparator to compare Docker API versions with only major and minor versions too

### Testing
- [x] Builds on Linux (`make release`)
- [x] Builds on Windows (`go build -out amazon-ecs-agent.exe ./agent`)
- [x] Unit tests on Linux (`make test`) pass
- [x] Unit tests on Windows (`go test -timeout=25s ./agent/...`) pass
- [x] Integration tests on Linux (`make run-integ-tests`) pass
- [x] Integration tests on Windows (`.\scripts\run-integ-tests.ps1`) pass
- [x] Functional tests on Linux (`make run-functional-tests`) pass
- [x] Functional tests on Windows (`.\scripts\run-functional-tests.ps1`) pass

New tests cover the changes: yes|no

### Description for the changelog
Supported Docker version clients from default client's min supported API and API version returned

### Licensing
This contribution is under the terms of the Apache 2.0 License: yes

### CODE REVIEW

**Comment by sharanyad** [comment]
This addresses https://github.com/aws/amazon-ecs-agent/issues/1010 required for https://github.com/aws/amazon-ecs-agent/pull/996

**Comment by aaithal** [comment]
@sharanyad can you please fix the unit test failure?

File: agent/engine/dockerclient/dockerclientfactory.go [hunk 1]
```diff
@@ -121,8 +135,54 @@ func findDockerVersions(endpoint string) map[DockerVersion]dockeriface.Client {
...
 		if err != nil {
 			log.Infof("Failed to ping with Docker version %s: %v", version, err)
+			continue
 		}
 		clients[version] = client
 	}
 	return clients
 }
+
+func findDockerVersionsfromMinMaxVersions(endpoint string) (map[DockerVersion]dockeriface.Client, bool) {
```
Comment (aaithal) [inline agent/engine/dockerclient/dockerclientfactory.go]: You're using the second parameter type (`bool`) to indicate the success/failure of this method, which means that returning an `error` instead is the better way to do that.

File: agent/engine/dockerclient/dockerclientfactory.go [hunk 2]
```diff
@@ -121,8 +135,54 @@ func findDockerVersions(endpoint string) map[DockerVersion]dockeriface.Client {
...
+	clients := make(map[DockerVersion]dockeriface.Client)
```
Comment (vsiddharth) [inline agent/engine/dockerclient/dockerclientfactory.go]: This line can be moved down below

File: agent/utils/compare_versions.go [hunk 3]
```diff
@@ -93,6 +96,9 @@ func (lhs Version) Matches(selector string) (bool, error) {
 
 func parseSemver(version string) (semver, error) {
 	var result semver
+	if ok := checkForNoPatch(version); ok {
```
[the remark this answers is not in the document]
Comment (sharanyad) [reply agent/utils/compare_versions.go]: done

File: agent/engine/dockerclient/dockerclientfactory.go (hunk 2 above)
Comment (sharanyad) [reply agent/engine/dockerclient/dockerclientfactory.go]: done

File: agent/engine/dockerclient/dockerclientfactory.go (hunk 1 above)
Comment (sharanyad) [reply agent/engine/dockerclient/dockerclientfactory.go]: I am not sure if it can be called a failure if min api version key is not found, nevertheless changed the return type to `error`

File: agent/engine/dockerclient/dockerclientfactory.go [hunk 4]
```diff
@@ -121,8 +135,54 @@ func findDockerVersions(endpoint string) map[DockerVersion]dockeriface.Client {
...
+	if err != nil {
+		log.Infof("Error while getting client version: %v", err)
+		return nil, false
+	}
+
+	// check if the docker.Env obj has MinAPIVersion key
+	if clientVersion.Exists(minAPIVersionKey) {
+		minAPIVersion := clientVersion.Get(minAPIVersionKey)
+		apiVersion := clientVersion.Get(apiVersionKey)
+		for _, version := range getKnownAPIVersions() {
```
[the remark this answers is not in the document]
Comment (sharanyad) [reply agent/engine/dockerclient/dockerclientfactory.go]: Sure.

File: agent/engine/dockerclient/dockerclientfactory.go [hunk 5]
```diff
@@ -14,13 +14,21 @@
...
 	"github.com/aws/amazon-ecs-agent/agent/engine/dockeriface"
+	"github.com/aws/amazon-ecs-agent/agent/utils"
 	log "github.com/cihub/seelog"
 	docker "github.com/fsouza/go-dockerclient"
+	"github.com/pkg/errors"
 )
 
+const (
+	// minDockerAPILinux is the min Docker API version supported by agent
+	minDockerAPILinux = Version_1_17
```
Comment (samuelkarp) [inline agent/engine/dockerclient/dockerclientfactory.go]: You need to specify this differently for Linux and Windows.  Windows won't respond at this API version.  You can change the constants in `versionsupport_windows.go` and `versionsupport_unix.go` to have the same name and then just use those.

File: agent/utils/compare_versions.go [hunk 6]
```diff
@@ -20,6 +20,9 @@ import (
 	"strings"
 )
 
+// zeroPatch is a string to append patch number zero if the major minor version lacks it
+const zeroPatch = ".0"
```
Comment (samuelkarp) [inline agent/utils/compare_versions.go]: Instead of doing this, I think I'd actually prefer implementing a different non-semver version comparator.  API versions from Docker are not semver and making them look like semver can just be a source of confusion.

File: agent/engine/dockerclient/dockerclientfactory.go [hunk 7]
```diff
@@ -121,8 +137,66 @@ func findDockerVersions(endpoint string) map[DockerVersion]dockeriface.Client {
...
+
+	clients := make(map[DockerVersion]dockeriface.Client)
+	// check if the docker.Env obj has MinAPIVersion key
+	if clientVersion.Exists(minAPIVersionKey) {
+		minAPIVersion := clientVersion.Get(minAPIVersionKey)
+		apiVersion := clientVersion.Get(apiVersionKey)
+		for _, version := range getKnownAPIVersions() {
+			dockerClient, err := getDockerClientForVersion(endpoint, string(version), minAPIVersion, apiVersion)
+			if err != nil {
+				log.Infof("Error getting client version: %v", err)
```
Comment (samuelkarp) [inline agent/engine/dockerclient/dockerclientfactory.go]: Is INFO the right level for this?  I think you actually want DEBUG and probably don't call it an error (since it is expected that some won't match).

File: agent/utils/compare_versions.go (hunk 6 above)
Comment (sharanyad) [reply agent/utils/compare_versions.go]: I just wanted to reuse the existing semver comparator due to time constraints for this fix. I agree this is not the right way, but I would prefer to add it to my list, come back and implement a non-semver comparator once this goes out.

Comment (samuelkarp) [reply agent/utils/compare_versions.go]: That's fair.  Want to add a TODO comment and possibly open an issue to track?

File: agent/engine/dockerclient/dockerclientfactory.go [hunk 8]
```diff
@@ -121,8 +137,66 @@ func findDockerVersions(endpoint string) map[DockerVersion]dockeriface.Client {
...
 		clients[version] = client
 	}
 	return clients
 }
+
+func findDockerVersionsfromMinMaxVersions(endpoint string) (map[DockerVersion]dockeriface.Client, error) {
+	// get a Docker client with the default supported version
+	client, err := newVersionedClient(endpoint, string(minDockerAPILinux))
+	if err != nil {
+		return nil, errors.Wrap(err, "Error while creating client")
```
Comment (vsiddharth) [inline agent/engine/dockerclient/dockerclientfactory.go]: Please add more context to the error and not use upper case characters in the message. 
The code base has plenty of examples using `errors.Wrap`

File: agent/engine/dockerclient/dockerclientfactory.go [hunk 9]
```diff
@@ -121,8 +137,66 @@ func findDockerVersions(endpoint string) map[DockerVersion]dockeriface.Client {
...
+	}
+
+	clientVersion, err := client.Version()
+	if err != nil {
+		return nil, errors.Wrap(err, "Error while getting client version")
```
Comment (vsiddharth) [inline agent/engine/dockerclient/dockerclientfactory.go]: same as above

File: agent/engine/dockerclient/dockerclientfactory.go [hunk 10]
```diff
@@ -121,8 +137,66 @@ func findDockerVersions(endpoint string) map[DockerVersion]dockeriface.Client {
...
+	endpoint string,
+	version string,
+	minAPIVersion string,
+	apiVersion string) (dockeriface.Client, error) {
+	lessThanMinCheck := "<"+minAPIVersion
+	moreThanMaxCheck := ">"+apiVersion
+	minVersionCheck, minErr := utils.Version(version).Matches(lessThanMinCheck)
+	maxVersionCheck, maxErr := utils.Version(version).Matches(moreThanMaxCheck)
+	if minErr != nil {
+		return nil, errors.Wrapf(minErr, "Error while comparing version %s with minAPIVersion %s", version, minAPIVersion)
```
Comment (vsiddharth) [inline agent/engine/dockerclient/dockerclientfactory.go]: same as above

File: agent/engine/dockerclient/dockerclientfactory.go [hunk 11]
```diff
@@ -121,8 +137,66 @@ func findDockerVersions(endpoint string) map[DockerVersion]dockeriface.Client {
...
+	}
+	if maxErr != nil {
+		return nil, errors.Wrapf(minErr, "Error while comparing version %s with apiVersion %s", version, apiVersion)
```
Comment (vsiddharth) [inline agent/engine/dockerclient/dockerclientfactory.go]: same as above

File: agent/engine/dockerclient/dockerclientfactory.go [hunk 12]
```diff
@@ -121,8 +137,66 @@ func findDockerVersions(endpoint string) map[DockerVersion]dockeriface.Client {
...
+	}
+	// do not add the version when it is less than min api version or greater
+	// than api version
+	if minVersionCheck || maxVersionCheck {
+		return nil, errors.Errorf("Min and API versions comparison check failed for version: %s", version)
```
Comment (vsiddharth) [inline agent/engine/dockerclient/dockerclientfactory.go]: same as above

File: agent/engine/dockerclient/dockerclientfactory.go [hunk 13]
```diff
@@ -121,8 +137,66 @@ func findDockerVersions(endpoint string) map[DockerVersion]dockeriface.Client {
...
+	}
+	client, err := newVersionedClient(endpoint, string(version))
+	if err != nil {
+		return nil, errors.Wrap(err, "Error while creating client")
```
Comment (vsiddharth) [inline agent/engine/dockerclient/dockerclientfactory.go]: same as above

File: agent/utils/compare_versions.go (hunk 6 above)
Comment (sharanyad) [reply agent/utils/compare_versions.go]: Sure, will do that.

File: agent/engine/dockerclient/dockerclientfactory.go (hunk 5 above)
Comment (vsiddharth) [inline agent/engine/dockerclient/dockerclientfactory.go]: I don't think this would work on both `linux` and `windows`. 
Please ensure that this change works well on all supported platforms.

**Comment by sharanyad** [comment]
@aithal @samuelkarp @vsiddharth Addressed the review comments. Could you please have a look again?

File: agent/engine/dockerclient/dockerclientfactory.go [hunk 14]
```diff
@@ -14,13 +14,19 @@
...
 	"github.com/aws/amazon-ecs-agent/agent/engine/dockeriface"
+	"github.com/aws/amazon-ecs-agent/agent/utils"
 	log "github.com/cihub/seelog"
 	docker "github.com/fsouza/go-dockerclient"
+	"github.com/pkg/errors"
 )
 
+const (
+	// minAPIVersionKey is the docker.Env key for min API version
+	minAPIVersionKey = "MinAPIVersion"
```
Comment (aaithal) [inline agent/engine/dockerclient/dockerclientfactory.go]: Can you also document here that this is supported in API versions `>=1.25` as per https://docs.docker.com/engine/api/version-history/#v125-api-changes ?

File: agent/engine/dockerclient/dockerclientfactory.go [hunk 15]
```diff
@@ -110,7 +116,15 @@ func (f *factory) getClient(version DockerVersion) (dockeriface.Client, error) {
 // findDockerVersions loops over all known API versions and finds which ones
 // are supported by the docker daemon on the host
 func findDockerVersions(endpoint string) map[DockerVersion]dockeriface.Client {
-	clients := make(map[DockerVersion]dockeriface.Client)
+	// if the client version returns a min version and api version, then use it to return
+	// Docker versions
```
Comment (aaithal) [inline agent/engine/dockerclient/dockerclientfactory.go]: Let's make this more explicit: `use it to return versions between MinAPIVersion and APIVerison`

File: agent/engine/dockerclient/dockerclientfactory.go [hunk 16]
```diff
@@ -121,8 +135,66 @@ func findDockerVersions(endpoint string) map[DockerVersion]dockeriface.Client {
...
+	}
+
+	clientVersion, err := client.Version()
+	if err != nil {
+		return nil, errors.Wrap(err, "min API version check: error while getting client version")
+	}
+
+	clients := make(map[DockerVersion]dockeriface.Client)
+	// check if the docker.Env obj has MinAPIVersion key
+	if clientVersion.Exists(minAPIVersionKey) {
```
Comment (aaithal) [inline agent/engine/dockerclient/dockerclientfactory.go]: let's refactor this a bit more:
```
if !clientVersion.Exists(minAPIVersionKey) {
 return ...
}

minAPIVersion := clientVersion.Get(minAPIVersionKey)
...
```

File: agent/engine/dockerclient/dockerclientfactory.go [hunk 17]
```diff
@@ -121,8 +135,66 @@ func findDockerVersions(endpoint string) map[DockerVersion]dockeriface.Client {
...
+}
+
+func getDockerClientForVersion(
+	endpoint string,
+	version string,
+	minAPIVersion string,
+	apiVersion string) (dockeriface.Client, error) {
+	lessThanMinCheck := "<"+minAPIVersion
+	moreThanMaxCheck := ">"+apiVersion
+	minVersionCheck, minErr := utils.Version(version).Matches(lessThanMinCheck)
```
Comment (aaithal) [inline agent/engine/dockerclient/dockerclientfactory.go]: I don't think you need multiple `error` variables here. If you stick to the policy of creating/initializing a variable only when needed, you'll do something like this:
```
minVersionCheck, err := utils.Version(version).Matches(lessThanMinCheck)
if err != nil {
  ...
}

maxVersionCheck, err := utils.Version(version).Matches(moreThanMaxCheck)
if err != nil {
 ...
}
if minVersionCheck || ...
```

File: agent/engine/dockerclient/dockerclientfactory.go [hunk 18]
```diff
@@ -110,7 +116,15 @@ func (f *factory) getClient(version DockerVersion) (dockeriface.Client, error) {
 // findDockerVersions loops over all known API versions and finds which ones
 // are supported by the docker daemon on the host
 func findDockerVersions(endpoint string) map[DockerVersion]dockeriface.Client {
-	clients := make(map[DockerVersion]dockeriface.Client)
+	// if the client version returns a min version and api version, then use it to return
+	// Docker versions
+	clients, err := findDockerVersionsfromMinMaxVersions(endpoint)
+	if err != nil {
+		log.Debugf("Unable to get Docker clients from min API version: %v", err)
```
Comment (aaithal) [inline agent/engine/dockerclient/dockerclientfactory.go]: I think this should be `Infof`

File: agent/engine/dockerclient/dockerclientfactory.go [hunk 19]
```diff
@@ -121,8 +135,66 @@ func findDockerVersions(endpoint string) map[DockerVersion]dockeriface.Client {
...
+
+	clients := make(map[DockerVersion]dockeriface.Client)
+	// check if the docker.Env obj has MinAPIVersion key
+	if clientVersion.Exists(minAPIVersionKey) {
+		minAPIVersion := clientVersion.Get(minAPIVersionKey)
+		apiVersion := clientVersion.Get(apiVersionKey)
+		for _, version := range getKnownAPIVersions() {
+			dockerClient, err := getDockerClientForVersion(endpoint, string(version), minAPIVersion, apiVersion)
+			if err != nil {
+				log.Debugf("Unable to get Docker client for version %s: %v", version, err)
```
Comment (aaithal) [inline agent/engine/dockerclient/dockerclientfactory.go]: I think this should be `Infof`

File: agent/utils/compare_versions.go [hunk 20]
```diff
@@ -40,7 +43,8 @@ type semver struct {
 // * <x.y.z -- Matches a version less than the selector version
 // * x.y.z,a.b.c -- Matches if the version matches either of the two selector versions
 func (lhs Version) Matches(selector string) (bool, error) {
-	lhsVersion, err := parseSemver(string(lhs))
+	lhsVer := appendIfNoPatch(string(lhs))
```
Comment (aaithal) [inline agent/utils/compare_versions.go]: Let's not do `appendIfNoPatch` anywhere in this method. Instead, do that in `getDockerClientForVersion` so that you don't have to keep doing that at multiple places here. Let's minimize changes to `Version` type as much as possible.

File: agent/engine/dockerclient/dockerclientfactory.go (hunk 19 above)
Comment (sharanyad) [reply agent/engine/dockerclient/dockerclientfactory.go]: It makes more sense to have it as `Debugf` to me, since we are just trying to obtain docker client for a particular version and it might not always be successful, which is expected. All the older versions' agent logs would contain these messages looking like errors in the beginning. So made it `Debugf`. What do you think?

File: agent/utils/compare_versions.go (hunk 20 above)
Comment (aaithal) [reply agent/utils/compare_versions.go]: If there are asynchronous failures for containers because of a specific versioned docker client  not being there, having these errors printed out during discovery helps us debug the issue. If we make these `Debug`, we'll lose these messages when logging at default `Info` level.

File: agent/engine/dockerclient/dockerclientfactory.go (hunk 19 above)
Comment (aaithal) [reply agent/engine/dockerclient/dockerclientfactory.go]: If there are asynchronous failures for containers because of a specific versioned docker client  not being there, having these errors printed out during discovery helps us debug the issue. If we make these `Debug`, we'll lose these messages when logging at default `Info` level.

Comment (samuelkarp) [reply agent/engine/dockerclient/dockerclientfactory.go]: @aaithal Theoretically...we shouldn't have async failures due to missing clients.  This discovery logic is used for capability registration, and if we don't register a capability there shouldn't be a task scheduled that requires an unavailable client.

Comment (aaithal) [reply agent/engine/dockerclient/dockerclientfactory.go]: @samuelkarp yeah, but bugs.. Since this is just a once time discovery thing, logging it at Info shouldn't contribute to too much noise IMO.

File: agent/engine/dockerclient/dockerclientfactory.go (hunk 17 above)
Comment (sharanyad) [reply agent/engine/dockerclient/dockerclientfactory.go]: Earlier was doing a check of the error messages together, didn't modify it later. changed now.

File: agent/engine/dockerclient/dockerclientfactory.go [hunk 21]
```diff
@@ -121,8 +139,70 @@ func findDockerVersions(endpoint string) map[DockerVersion]dockeriface.Client {
...
+		return nil, errors.Wrapf(err, "error while comparing version %s with minAPIVersion %s", version, minAPIVersion)
+	}
+	maxVersionCheck, err := utils.Version(version).Matches(moreThanMaxCheck)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error while comparing version %s with apiVersion %s", version, apiVersion)
+	}
+	// do not add the version when it is less than min api version or greater
+	// than api version
+	if minVersionCheck || maxVersionCheck {
+		return nil, errors.Errorf("min and API versions comparison check failed for version: %s", version)
```
Comment (aaithal) [inline agent/engine/dockerclient/dockerclientfactory.go]: Can you rephrase this? `"version detection using MinAPIVersion: unsupported version: %s"`

File: agent/engine/dockerclient/dockerclientfactory.go [hunk 22]
```diff
@@ -121,8 +139,70 @@ func findDockerVersions(endpoint string) map[DockerVersion]dockeriface.Client {
...
+	version += zeroPatch
+	lessThanMinCheck := "<" + minAPIVersion + zeroPatch
+	moreThanMaxCheck := ">" + apiVersion + zeroPatch
+	minVersionCheck, err := utils.Version(version).Matches(lessThanMinCheck)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error while comparing version %s with minAPIVersion %s", version, minAPIVersion)
+	}
+	maxVersionCheck, err := utils.Version(version).Matches(moreThanMaxCheck)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error while comparing version %s with apiVersion %s", version, apiVersion)
```
Comment (aaithal) [inline agent/engine/dockerclient/dockerclientfactory.go]: Can you rephrase this? `"version detection using MinAPIVersion: unable to get max version: %s"`

File: agent/engine/dockerclient/dockerclientfactory.go [hunk 23]
```diff
@@ -121,8 +139,70 @@ func findDockerVersions(endpoint string) map[DockerVersion]dockeriface.Client {
...
+	apiVersion string) (dockeriface.Client, error) {
+	// Adding patch number zero to Docker versions to reuse the existing semver
+	// comparator
+	// TODO: remove this logic later when non-semver comparator is implemented
+	version += zeroPatch
+	lessThanMinCheck := "<" + minAPIVersion + zeroPatch
+	moreThanMaxCheck := ">" + apiVersion + zeroPatch
+	minVersionCheck, err := utils.Version(version).Matches(lessThanMinCheck)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error while comparing version %s with minAPIVersion %s", version, minAPIVersion)
```
Comment (aaithal) [inline agent/engine/dockerclient/dockerclientfactory.go]: Can you rephrase this? `"version detection using MinAPIVersion: unable to get min version: %s"`

File: agent/engine/dockerclient/dockerclientfactory.go [hunk 24]
```diff
@@ -121,8 +139,70 @@ func findDockerVersions(endpoint string) map[DockerVersion]dockeriface.Client {
...
+		return nil, errors.Wrapf(err, "error while comparing version %s with apiVersion %s", version, apiVersion)
+	}
+	// do not add the version when it is less than min api version or greater
+	// than api version
+	if minVersionCheck || maxVersionCheck {
+		return nil, errors.Errorf("min and API versions comparison check failed for version: %s", version)
+	}
+	client, err := newVersionedClient(endpoint, string(version))
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to create Docker client for version: %s", version)
```
Comment (aaithal) [inline agent/engine/dockerclient/dockerclientfactory.go]: Can you rephrase this? `"version detection using MinAPIVersion: unable to create Docker client for version: %s"`

File: agent/utils/compare_versions_test.go [hunk 25]
```diff
@@ -185,6 +185,11 @@ func TestVersionMatches(t *testing.T) {
 			selector:       ">=1.5.0",
 			expectedOutput: true,
 		},
+		{
+			version:	"1.17",
```
Comment (aaithal) [inline agent/utils/compare_versions_test.go]: Shouldn't you move this to `engine/dockerclient` package as well?

Comment (sharanyad) [reply agent/utils/compare_versions_test.go]: yeah, this test should not be here, will remove it. There is a general test added in `engine/dockerclient` for getting Docker versions from `minAPIVersion`.

File: agent/engine/dockerclient/dockerclientfactory.go [hunk 26]
```diff
@@ -121,8 +139,70 @@ func findDockerVersions(endpoint string) map[DockerVersion]dockeriface.Client {
...
+		return nil, errors.Wrap(err, "min API version check: error while getting client version")
+	}
+
+	clients := make(map[DockerVersion]dockeriface.Client)
+	// check if the docker.Env obj has MinAPIVersion key
+	if !clientVersion.Exists(minAPIVersionKey) {
+		return nil, errors.New("min API version not found in client version")
+	}
+	minAPIVersion := clientVersion.Get(minAPIVersionKey)
+	apiVersion := clientVersion.Get(apiVersionKey)
```
Comment (richardpen) [inline agent/engine/dockerclient/dockerclientfactory.go]: Can you add a check here for `minAPIVersion` or `apiVersion` is empty?

File: agent/engine/dockerclient/dockerclientfactory.go [hunk 27]
```diff
@@ -110,7 +120,15 @@ func (f *factory) getClient(version DockerVersion) (dockeriface.Client, error) {
 // findDockerVersions loops over all known API versions and finds which ones
 // are supported by the docker daemon on the host
 func findDockerVersions(endpoint string) map[DockerVersion]dockeriface.Client {
-	clients := make(map[DockerVersion]dockeriface.Client)
+	// if the client version returns a MinAPIVersion and APIVersion, then use it to return
+	// all the Docker clients between MinAPIVersion and APIVersion
+	clients, err := findDockerVersionsfromMinMaxVersions(endpoint)
```
Comment (petderek) [inline agent/engine/dockerclient/dockerclientfactory.go]: Two concerns here. The first is that this is going to behave differently on windows. Windows only supports API version 1.24 or greater. If we put this check here, it will skip over the capability logic that ensures tasks are schedule-able to windows.

File: agent/engine/dockerclient/dockerclientfactory.go [hunk 28]
```diff
@@ -110,7 +120,15 @@ func (f *factory) getClient(version DockerVersion) (dockeriface.Client, error) {
...
+	if err != nil {
+		log.Infof("Unable to get Docker clients from min API version: %v", err)
+	} else {
+		return clients
```
Comment (petderek) [inline agent/engine/dockerclient/dockerclientfactory.go]: The second concern is tangential. Putting this here puts a hard branch in our detection logic, and I think we should consider composing it differently.

```golang
// As written, the code boils down to this
clients, err := findDockerVersionsOneWay()
if err == nil {
    return clients
}
clients, err := findDockerVersionsADifferentWay()
```
I'd rather see us handle the detection in the same place using existing logic instead of branching like this. One suggestion would be to look at the `version_support_{platform}` files and the Min version variable. You could write a function that does something like this:

```golang
 // Use either the docker minimum API version or rely on the version set in version_support_{platform}.go
func getMinVersion() DockerVersion {
    if functionThatChecksIfDockerHasMinAPIParam() {
        return functionThatExtractsMinAPIFromDockerAPI()
    } else {
        return minDockerAPIVersion
    }
}

File: agent/engine/dockerclient/dockerclientfactory.go (hunk 27 above)
Comment (aaithal) [reply agent/engine/dockerclient/dockerclientfactory.go]: Yeah, I agree. I missed this in my review. Good call!

File: agent/engine/dockerclient/dockerclientfactory.go (hunk 28 above)
Comment (petderek) [reply agent/engine/dockerclient/dockerclientfactory.go]: Oh &mdash; You'd want to write a function to check for the max version, too :)

File: agent/engine/dockerclient/dockerclientfactory.go (hunk 27 above)
Comment (sharanyad) [reply agent/engine/dockerclient/dockerclientfactory.go]: According to current logic for `FindSupportedAPIVersions`, if minimum version client `1.17` is not supported in Windows, the tasks should anyway not be scheduled on Windows right?

File: agent/engine/dockerclient/dockerclientfactory.go (hunk 28 above)
Comment (sharanyad) [reply agent/engine/dockerclient/dockerclientfactory.go]: What would be the default max version here be?

File: agent/engine/dockerclient/dockerclientfactory.go (hunk 27 above)
Comment (samuelkarp) [reply agent/engine/dockerclient/dockerclientfactory.go]: No, not quite right.  We register capabilities for API versions below 1.24 on Windows because we still want tasks that use features in those APIs to work on Windows too.  Our capability logic is pretty naive right now: we map a set of features to the API version in which they were introduced and then use that as an indicator of the minimum version of Docker they'll run on.  Because Docker on Windows still supports (many of) the features that work on older versions of Docker, this was an inexpensive way to get that support without rebuilding our capability logic.

I agree that this is confusing and non-obvious.  I think long-term we probably want to revisit this logic and see if we can improve it, given that we now know different things from when we did when it was designed.  However, we haven't done that yet and we still need to register capabilities for older API versions on Windows.

File: agent/engine/dockerclient/dockerclientfactory.go [hunk 29]
```diff
@@ -110,19 +120,72 @@ func (f *factory) getClient(version DockerVersion) (dockeriface.Client, error) {
 // findDockerVersions loops over all known API versions and finds which ones
 // are supported by the docker daemon on the host
 func findDockerVersions(endpoint string) map[DockerVersion]dockeriface.Client {
+	// if the client version returns a MinAPIVersion and APIVersion, then use it to return
+	// all the Docker clients between MinAPIVersion and APIVersion, else try pinging
+	// the clients in getKnownAPIVersions
+	minAPIVersion := ""
```
Comment (vsiddharth) [inline agent/engine/dockerclient/dockerclientfactory.go]: This can be composed more idiomatically
`var minAPIVersion, apiVersion string`

File: agent/engine/dockerclient/dockerclientfactory.go [hunk 30]
```diff
@@ -110,19 +120,72 @@ func (f *factory) getClient(version DockerVersion) (dockeriface.Client, error) {
...
+		lessThanMinCheck := "<" + minAPIVersion + zeroPatch
+		moreThanMaxCheck := ">" + apiVersion + zeroPatch
+		minVersionCheck, err := utils.Version(versionWithPatch).Matches(lessThanMinCheck)
+		if err != nil {
+			return nil, errors.Wrapf(err, "version detection using MinAPIVersion: unable to get min version: %s", minAPIVersion)
+		}
+		maxVersionCheck, err := utils.Version(versionWithPatch).Matches(moreThanMaxCheck)
 		if err != nil {
-			log.Infof("Failed to ping with Docker version %s: %v", version, err)
+			return nil, errors.Wrapf(err, "version detection using MinAPIVersion: unable to get max version: %s", apiVersion)
```
Comment (vsiddharth) [inline agent/engine/dockerclient/dockerclientfactory.go]: Should this be `MinAPIVersion`?

File: agent/engine/dockerclient/dockerclientfactory.go [hunk 31]
```diff
@@ -110,19 +120,72 @@ func (f *factory) getClient(version DockerVersion) (dockeriface.Client, error) {
...
+		}
+		// do not add the version when it is less than min api version or greater
+		// than api version
+		if minVersionCheck || maxVersionCheck {
+			return nil, errors.Errorf("version detection using MinAPIVersion: unsupported version: %s", version)
```
Comment (vsiddharth) [inline agent/engine/dockerclient/dockerclientfactory.go]: Should this be `MinAPIVersion` again ?

File: agent/engine/dockerclient/dockerclientfactory.go (hunk 27 above)
Comment (sharanyad) [reply agent/engine/dockerclient/dockerclientfactory.go]: Got it. 
From reading today's code, in `findDockerVersions` for windows, we iterate through the `KnownAPIVersions` from 1.17. Wouldn't trying to create a client with version 1.17 or trying to ping, fail in Docker Windows? Eventually, those versions would not be added to the  list of clients that Docker supports and hence the capability for 1.17 would not be added? 
https://github.com/aws/amazon-ecs-agent/blob/master/agent/engine/dockerclient/dockerclientfactory.go#L115

Comment (samuelkarp) [reply agent/engine/dockerclient/dockerclientfactory.go]: We don't actually want to create a client at 1.17 on Windows or even to attempt to ping it; we know that it'll fail (except on Docker 1.12 beta).  I'm talking strictly about capability registration here in that we still need to register those capabilities (and then override the incoming API version with the minimum of 1.24).

Comment (sharanyad) [reply agent/engine/dockerclient/dockerclientfactory.go]: But yes, I agree we should revisit this logic sometime and adapt it for windows.

File: agent/engine/dockerclient/dockerclientfactory.go (hunk 31 above)
Comment (sharanyad) [reply agent/engine/dockerclient/dockerclientfactory.go]: No, we are simply logging the version from from `knownVersions` , which does not fall in the range `[minAPIVersion, apiVersion]`

File: agent/engine/dockerclient/dockerclientfactory.go (hunk 30 above)
Comment (sharanyad) [reply agent/engine/dockerclient/dockerclientfactory.go]: No, this means something was wrong with `apiVersion` and the comparison check failed.

File: agent/engine/dockerclient/dockerclientfactory.go [hunk 32]
```diff
@@ -110,19 +120,71 @@ func (f *factory) getClient(version DockerVersion) (dockeriface.Client, error) {
 // findDockerVersions loops over all known API versions and finds which ones
 // are supported by the docker daemon on the host
 func findDockerVersions(endpoint string) map[DockerVersion]dockeriface.Client {
+	// if the client version returns a MinAPIVersion and APIVersion, then use it to return
+	// all the Docker clients between MinAPIVersion and APIVersion, else try pinging
+	// the clients in getKnownAPIVersions
+	var minAPIVersion, apiVersion string
+	// get a Docker client with the default supported version
+	client, err := newVersionedClient(endpoint, string(minDockerAPIVersion))
+	if err == nil {
```
Comment (samuelkarp) [inline agent/engine/dockerclient/dockerclientfactory.go]: I would find this easier to read if it were refactored a bit and the conditions were inverted, something like this:

```go
func findDockerVersionsFromAPI(endpoint string) (int, int, bool) {
	client, err := newVersionedClient(endpoint, string(minDockerAPIVersion))
	if err != nil {
		seelog.Debug("Failed to query the version API", err)
		return "", "", false
	}

	clientVersion, err := client.Version()
	if err != nil {
		seelog.Debug("Failed to extract version", err)
		return "", "", false
	}

	if !clientVersion.Exists(minAPIVersionKey) || !clientVersion.Exists(apiVersionKey) {
		seelog.Debug("Failed to extract minimum and maximum version")
		return "", "", false
	}

	return clientVersion.Get(minAPIVersionKey), clientVersion.Get(apiVersionKey), true
}
```

This way the error conditions are more clear and there's less to keep in mind as you're reading through.
